narwhals/dataframe.py

```diff
@@ -137,10 +138,18 @@ def head(self, n: int) -> Self:
     def tail(self, n: int) -> Self:
         return self._from_compliant_dataframe(self._compliant_frame.tail(n))
 
-    def drop(self, *columns: str | Iterable[str]) -> Self:
-        return self._from_compliant_dataframe(
-            self._compliant_frame.drop(*flatten(columns))
-        )
+    def drop(self, *columns: Iterable[str], strict: bool) -> Self:
+        cols = set(self.collect_schema().names())
```

Comment:
this is the part I'm a little unsure about

I'm not sure we should be triggering `collect_schema` for Polars LazyFrame - I'll check what they're doing

Reply:
We can let polars do what polars does natively.
I can factor out the parsing into a utils function and call it downstream instead of in `BaseFrame`.

Reply:
thanks! sure, but at the moment we still reach this `collect_schema` call for Polars, right? that's the part i think we should avoid

Reply:
yes at the moment we do. I will rework this PR a bit to avoid that in the versions that polars natively supports `.drop(..., strict=...)`

Reply:
maybe we can just avoid it completely for Polars, and note in the docstring that for Polars<1 dropping non-existent columns silently passes?

Reply:
Whops! I was doing the refactor and just read this.

In polars pre v1 passing a column not present in the data just ends up doing nothing. I wonder how v1 does it as it has to know at least the column names, so I would expect implicitly to collect those?!

Reply:
> so I would expect implicitly to collect those?!

yup, but only a `LazyFrame.collect` time, not at `LazyFrame.drop` time
```
In [8]: df = pl.LazyFrame({'a': [1,2,3]})

In [9]: df.drop('b')
Out[9]: <LazyFrame at 0x7FE016F33510>

In [10]: df.drop('b').collect()
---------------------------------------------------------------------------
ColumnNotFoundError                       Traceback (most recent call last)
Cell In[10], line 1
----> 1 df.drop('b').collect()

File ~/scratch/.venv/lib/python3.11/site-packages/polars/lazyframe/frame.py:1942, in LazyFrame.collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, cluster_with_columns, no_optimization, streaming, background, _eager, **_kwargs)
   1939 # Only for testing purposes atm.
   1940 callback = _kwargs.get("post_opt_callback")
-> 1942 return wrap_df(ldf.collect(callback))

ColumnNotFoundError: b
```

Reply:
Just to make sure to be on the same page before making other changes:

- Eager and lazy post v1 we let polars do its thing
- Eager pre v1, we can explicitly check column names as per other backend
- Lazy pre v1, ignore the strict argument and raise a warning saying something along the line of "please go eager here" if passed as True?

Reply:
thanks for checking, i'd say:


- _Eager and lazy post v1 we let polars do its thing_: agree
- _Eager pre v1, we can explicitly check column names as per other backend_ agree
- _Lazy pre v1, ignore the strict argument and raise a warning saying something along the line of "please go eager here" if passed as True?_ I was thinking more: just let Polars do its thinking, but leave a note in the docs about the pre-v1 behaviour difference. So, a docs note, not a runtime warning (which might be what you meant, just emphasising to be sure)

Reply:
I meant a runtime warning, especially because `strict=True` is the default, and that's the thing we ignore

Reply:
Reckon a runtime warning risks being annoying, as it may be unnecessary? And going eager might represent a significant performance degradation?